### ISSUE TITLE
Feature request: sample_weight option for log_prob

### ISSUE THREAD

**Issue by tom-andersson** [tensorflow/probability#1461]
I'm training a TensorFlow deep learning model with a `y_pred` that is a `tfp.distributions.Distribution` output. I want the loss function to be the negative log likelihood of the observations: `-y_pred.log_prob(y_true)`. However, in my application it is crucial to mask out some values from the `log_prob` summation using a `sample_weight` tensor of 0s and 1s whose shape matches `y_true`. I couldn't find any information on this on the `tensorflow-probability` documentation. Would it make sense to have a feature that allows for `y_pred.log_prob(y_true, sample_weight)`? Or is there a better approach? Is there a workaround to get the desired behaviour?

This issue relates to [this SO post](https://stackoverflow.com/questions/69886735/tensorflow-probability-how-to-do-sample-weighting-with-log-prob-loss-function), but swift feedback would be extremely useful (and it seems like there is a missing feature or documentation), hence raising this as an issue as well.

**Comment by SiegeLordEx** [comment]
You can use `tfd.Masked(y_pred, sample_weight).log_prob(y_true)`.

**Comment by tom-andersson** [comment]
Thanks very much @SiegeLordEx, I hadn't seen `tfd.Masked`! While this solves the issue I described in my OP, this wouldn't allow for weighting different samples by non-binary values, which is my actual problem (sorry!). I want to upweight some elements of the loss function by having values > 1 in the `sample_weight` tensor. It looks like `tfd.Masked` only works for boolean (binary) masks.

I've found an alternative solution using pure TensorFlow. My model outputs a `tfp.Independent` distribution, which means all the individual distributions in the `event_shape` are treated as a single random variable, meaning the `log_prob` is the scalar sum of all the individual `log_prob`s (not allowing for sample weighting to occur). One can get the tensor of `log_prob` values by accessing the `.distribution` attribute of the `tfp.Independent` object to absorb all the `event_shape` intro `batch_shape` and treat each element of the loss as an independent random variable. By writing a loss function that inherits from `tf.keras.losses.Loss`, the resulting weighted tensor is implicitly reduced, returning the weighted mean of `log_prob` values rather than the sum:

```
class NLL(tf.keras.losses.Loss):
    ''' Custom keras loss/metric for weighted negative log likelihood '''

    def __call__(self, y_true, y_pred, sample_weight=None):

        # This tensor is implicitly reduced by TensorFlow
        #     by taking the mean over all weighted elements
        return -y_pred.distribution.log_prob(y_true) * sample_weight
```
